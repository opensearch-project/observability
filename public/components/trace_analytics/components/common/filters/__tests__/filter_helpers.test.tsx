/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {
  getFilterFields,
  getInvertedOperator,
  getOperatorOptions,
  getValidFilterFields,
  getValueComponent,
} from '../filter_helpers';

describe('Filter helper functions', () => {
  configure({ adapter: new Adapter() });

  it('returns fields by page', () => {
    const fields = getFilterFields('dashboard');
    expect(fields).toEqual([
      'traceGroup',
      'serviceName',
      'error',
      'status.message',
      'latency',
    ]);
  });

  it('returns valid fields by page', () => {
    const dashboardFields = getValidFilterFields('dashboard');
    const servicesFields = getValidFilterFields('services');
    expect(dashboardFields).toEqual([
      'traceGroup',
      'serviceName',
      'error',
      'status.message',
      'latency',
      'Latency percentile within trace group',
    ]);
    expect(servicesFields).toEqual([
      'traceGroup',
      'serviceName',
      'error',
      'status.message',
      'latency',
    ]);
  });

  it('returns inverted operators', () => {
    const invertedBetween = getInvertedOperator('is between', true);
    const invertedExist = getInvertedOperator('exists', true);
    const invertedIs = getInvertedOperator('is', true);
    expect(invertedBetween).toEqual('is not between');
    expect(invertedExist).toEqual('does not exist');
    expect(invertedIs).toEqual('is not');
  });

  it('returns operator options by field', () => {
    const options = getOperatorOptions('durationInNanos');
    expect(options).toEqual([
      {
        label: 'is',
      },
      {
        label: 'is not',
      },
      {
        label: 'is between',
      },
      {
        label: 'is not between',
      },
      {
        label: 'exists',
      },
      {
        label: 'does not exist',
      },
    ]);
  });

  it('renders textfield filter', () => {
    const setValue = jest.fn((v) => {});
    const wrapper = mount(getValueComponent('serviceName', 'is', 0, setValue));
    expect(wrapper).toMatchSnapshot();

    wrapper.find('input').simulate('change', { target: { value: '100' } });
    expect(setValue).toBeCalledWith('100');
  });

  it('renders range field filter', () => {
    const setValue = jest.fn((v) => {});
    const wrapper = mount(
      getValueComponent('latency', 'is not between', { from: '0', to: '100' }, setValue)
    );
    expect(wrapper).toMatchSnapshot();

    wrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { value: '50' } });
    expect(setValue).toBeCalledWith({ from: '50', to: '100' });

    wrapper
      .find('input')
      .at(1)
      .simulate('change', { target: { value: '200' } });
    expect(setValue).toBeCalledWith({ from: '0', to: '200' });
  });
});
