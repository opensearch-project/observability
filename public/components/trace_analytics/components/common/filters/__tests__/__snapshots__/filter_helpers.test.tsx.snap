// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter helper functions renders range field filter 1`] = `
Array [
  <EuiSpacer
    size="s"
  >
    <div
      className="euiSpacer euiSpacer--s"
    />
  </EuiSpacer>,
  <EuiFormRow
    describedByIds={Array []}
    display="row"
    fullWidth={false}
    hasChildLabel={true}
    hasEmptyLabelSpace={false}
    label="Value"
    labelType="label"
  >
    <div
      className="euiFormRow"
      id="random_html_id-row"
    >
      <div
        className="euiFormRow__labelWrapper"
      >
        <EuiFormLabel
          className="euiFormRow__label"
          htmlFor="random_html_id"
          isFocused={false}
          type="label"
        >
          <label
            className="euiFormLabel euiFormRow__label"
            htmlFor="random_html_id"
          >
            Value
          </label>
        </EuiFormLabel>
      </div>
      <div
        className="euiFormRow__fieldWrapper"
      >
        <EuiFormControlLayoutDelimited
          endControl={
            <input
              className="euiFieldText"
              onChange={[Function]}
              placeholder="End of range"
              type="string"
              value="100"
            />
          }
          id="random_html_id"
          onBlur={[Function]}
          onFocus={[Function]}
          startControl={
            <input
              className="euiFieldText"
              onChange={[Function]}
              placeholder="Start of range"
              type="string"
              value="0"
            />
          }
        >
          <EuiFormControlLayout
            className="euiFormControlLayoutDelimited"
            id="random_html_id"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <div
              className="euiFormControlLayout euiFormControlLayoutDelimited"
              id="random_html_id"
              onBlur={[Function]}
              onFocus={[Function]}
            >
              <div
                className="euiFormControlLayout__childrenWrapper"
              >
                <input
                  className="euiFieldText euiFormControlLayoutDelimited__input"
                  onChange={[Function]}
                  placeholder="Start of range"
                  type="string"
                  value="0"
                />
                <EuiText
                  className="euiFormControlLayoutDelimited__delimeter"
                  color="subdued"
                  size="s"
                >
                  <div
                    className="euiText euiText--small euiFormControlLayoutDelimited__delimeter"
                  >
                    <EuiTextColor
                      color="subdued"
                      component="div"
                    >
                      <div
                        className="euiTextColor euiTextColor--subdued"
                      >
                        â†’
                      </div>
                    </EuiTextColor>
                  </div>
                </EuiText>
                <input
                  className="euiFieldText euiFormControlLayoutDelimited__input"
                  onChange={[Function]}
                  placeholder="End of range"
                  type="string"
                  value="100"
                />
                <EuiFormControlLayoutIcons />
              </div>
            </div>
          </EuiFormControlLayout>
        </EuiFormControlLayoutDelimited>
      </div>
    </div>
  </EuiFormRow>,
]
`;

exports[`Filter helper functions renders textfield filter 1`] = `
Array [
  <EuiSpacer
    size="s"
  >
    <div
      className="euiSpacer euiSpacer--s"
    />
  </EuiSpacer>,
  <EuiFormRow
    describedByIds={Array []}
    display="row"
    fullWidth={false}
    hasChildLabel={true}
    hasEmptyLabelSpace={false}
    label="Value"
    labelType="label"
  >
    <div
      className="euiFormRow"
      id="random_html_id-row"
    >
      <div
        className="euiFormRow__labelWrapper"
      >
        <EuiFormLabel
          className="euiFormRow__label"
          htmlFor="random_html_id"
          isFocused={false}
          type="label"
        >
          <label
            className="euiFormLabel euiFormRow__label"
            htmlFor="random_html_id"
          >
            Value
          </label>
        </EuiFormLabel>
      </div>
      <div
        className="euiFormRow__fieldWrapper"
      >
        <EuiFieldText
          id="random_html_id"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Enter a value"
          value={0}
        >
          <EuiFormControlLayout
            fullWidth={false}
            inputId="random_html_id"
          >
            <div
              className="euiFormControlLayout"
            >
              <div
                className="euiFormControlLayout__childrenWrapper"
              >
                <EuiValidatableControl>
                  <input
                    className="euiFieldText"
                    id="random_html_id"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Enter a value"
                    type="text"
                    value={0}
                  />
                </EuiValidatableControl>
                <EuiFormControlLayoutIcons />
              </div>
            </div>
          </EuiFormControlLayout>
        </EuiFieldText>
      </div>
    </div>
  </EuiFormRow>,
]
`;
